core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala

```diff
@@ -181,17 +181,33 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
       }
 
       try {
+        // get a server socket so that the launched daemon can tell us its server port
+        val serverSocket = new ServerSocket(0, 0, InetAddress.getByAddress(Array(127, 0, 0, 1)))
+        val serverPort = serverSocket.getLocalPort
+
         // Create and start the daemon
-        val pb = new ProcessBuilder(Arrays.asList(pythonExec, "-m", daemonModule))
+        val pb = new ProcessBuilder(Arrays.asList(pythonExec, "-m", daemonModule,
+          serverPort.toString))
         val workerEnv = pb.environment()
         workerEnv.putAll(envVars.asJava)
         workerEnv.put("PYTHONPATH", pythonPath)
         // This is equivalent to setting the -u flag; we use it because ipython doesn't support -u:
         workerEnv.put("PYTHONUNBUFFERED", "YES")
         daemon = pb.start()
 
+        // get the local port of the daemon's server socket,
+        // but don't wait forever for the daemon to connect
+        serverSocket.setSoTimeout(10000)
+        val socketToDaemon = serverSocket.accept()
```

Comment:
So, the problem with TCP sockets is: what if some other malicious process connects here instead of the process you expect to?

You can add an auth token and expose it to the child in an env variable, but that complicates the code a bit more. Sometimes I wish we could create proper pipes between processes in Java...

(You also probably want some `try..finally` error handling to close the server socket, at least.)

Reply:
>You also probably want some try..finally error handling to close the server socket, at least.

I take this as we don't want some serverSocket.close() error to kill the user's action. That makes sense to me, just checking.

Reply:
Not to kill the action, but just to make sure sockets are closed when they should.

Reply:
>You can add an auth token and expose it to the child in an env variable

How about a randomly generated integer passed to the daemon in the environment and then expected back as the first bit of data in the connection?

Reply:
Yeah, that's a reasonable way.

Reply:
>Not to kill the action, but just to make sure sockets are closed when they should.

Oh, of course, got it.